src/Microsoft.TestPlatform.CrossPlatEngine/Client/ProxyOperationManager.cs

```diff
@@ -82,16 +87,35 @@ public virtual void SetupChannel(IEnumerable<string> sources)
 
                 if (testHostStartInfo != null)
                 {
-                    // Monitor testhost exit.
-                    testHostStartInfo.ExitCallback = (process) =>
+                    // Monitor testhost error callbacks.
+                    testHostStartInfo.ErrorReceivedCallback = (process, data) =>
                     {
-                        if (process.ExitCode != 0)
+                        if(data != null)
                         {
-                            testHostProcessStdError = process.StandardError.ReadToEnd();
-                            EqtTrace.Error("Test host exited with error: {0}", testHostProcessStdError);
+                            //if incoming data stream is huge empty entire testError stream, & limit data stream to MaxCapacity
+                            if (data.Length > testHostProcessStdError.MaxCapacity)
+                            {
+                                testHostProcessStdError.Remove(0, testHostProcessStdError.Length);
+                                data = data.Substring(data.Length - testHostProcessStdError.MaxCapacity);
+                            }
+
+                            //remove only what is required, from begining of error stream
+                            else
+                            {
+                                int required = data.Length + testHostProcessStdError.Length - testHostProcessStdError.MaxCapacity;
+                                if (required > 0)
+                                {
+                                    testHostProcessStdError.Remove(0, required);
+                                }
+                            }
+
+                            testHostProcessStdError.Append(data);
                         }
-
-                        this.RequestSender.OnClientProcessExit(testHostProcessStdError);
+                        if (process.HasExited && process.ExitCode != 0)
```

Comment:
`process.HasExited` can throw InvalidOperationException if no process associated with object,  `process.ExitCode` can throw IOE if the process has not exited. is there a guaranty process.HasExited doesn't throw in callback?